File: app/src/main/java/org/thoughtcrime/securesms/util/UsernameUtil.java
package org.thoughtcrime.securesms.util;

import android.text.TextUtils;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.annotation.WorkerThread;

import org.signal.core.util.logging.Log;
import org.signal.libsignal.usernames.BaseUsernameException;
import org.signal.libsignal.usernames.Username;
import org.thoughtcrime.securesms.database.SignalDatabase;
import org.thoughtcrime.securesms.dependencies.ApplicationDependencies;
import org.thoughtcrime.securesms.recipients.Recipient;
import org.thoughtcrime.securesms.recipients.RecipientId;
import org.whispersystems.signalservice.api.push.ServiceId.ACI;
import org.whispersystems.signalservice.api.push.ServiceId;
import org.whispersystems.util.Base64UrlSafe;

import java.io.IOException;
import java.nio.charset.StandardCharsets;
import java.util.Locale;
import java.util.Optional;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

public class UsernameUtil {

  private static final String TAG = Log.tag(UsernameUtil.class);

  public static final int MIN_LENGTH = 3;
  public static final int MAX_LENGTH = 32;

  private static final Pattern FULL_PATTERN        = Pattern.compile(String.format(Locale.US, "^[a-zA-Z_][a-zA-Z0-9_]{%d,%d}$", MIN_LENGTH - 1, MAX_LENGTH - 1), Pattern.CASE_INSENSITIVE);
  private static final Pattern DIGIT_START_PATTERN = Pattern.compile("^[0-9].*$");
  private static final Pattern URL_PATTERN         = Pattern.compile("(https://)?signal.me/#u/([a-zA-Z0-9+/]*={0,2})");


  private static final String BASE_URL_SCHEMELESS = "signal.me/#u/";
  private static final String BASE_URL            = "https://" + BASE_URL_SCHEMELESS;

  public static boolean isValidUsernameForSearch(@Nullable String value) {
    return !TextUtils.isEmpty(value) && !DIGIT_START_PATTERN.matcher(value).matches();
  }

  public static Optional<InvalidReason> checkUsername(@Nullable String value) {
    if (value == null) {
      return Optional.of(InvalidReason.TOO_SHORT);
    } else if (value.length() < MIN_LENGTH) {
      return Optional.of(InvalidReason.TOO_SHORT);
    } else if (value.length() > MAX_LENGTH) {
      return Optional.of(InvalidReason.TOO_LONG);
    } else if (DIGIT_START_PATTERN.matcher(value).matches()) {
      return Optional.of(InvalidReason.STARTS_WITH_NUMBER);
    } else if (!FULL_PATTERN.matcher(value).matches()) {
      return Optional.of(InvalidReason.INVALID_CHARACTERS);
    } else {
      return Optional.empty();
    }
  }

  @WorkerThread
  public static @NonNull Optional<ServiceId> fetchAciForUsername(@NonNull String username) {
    Optional<RecipientId> localId = SignalDatabase.recipients().getByUsername(username);

    if (localId.isPresent()) {
      Recipient recipient = Recipient.resolved(localId.get());

      if (recipient.getServiceId().isPresent()) {
        Log.i(TAG, "Found username locally -- using associated UUID.");
        return recipient.getServiceId();
      } else {
        Log.w(TAG, "Found username locally, but it had no associated UUID! Clearing it.");
        SignalDatabase.recipients().clearUsernameIfExists(username);
      }
    }

    Log.d(TAG, "No local user with this username. Searching remotely.");
    try {
      return fetchAciForUsernameHash(Base64UrlSafe.encodeBytesWithoutPadding(Username.hash(username)));
    } catch (BaseUsernameException e) {
      return Optional.empty();
    }
  }

  /**
   * Hashes a username to a url-safe base64 string.
   * @throws BaseUsernameException If the username is invalid and un-hashable.
   */
  public static String hashUsernameToBase64(String username) throws BaseUsernameException {
    return Base64UrlSafe.encodeBytesWithoutPadding(Username.hash(username));
  }

  @WorkerThread
  public static @NonNull Optional<ServiceId> fetchAciForUsernameHash(@NonNull String base64UrlSafeEncodedUsernameHash) {
    try {
      ACI aci = ApplicationDependencies.getSignalServiceAccountManager()
                                       .getAciByUsernameHash(base64UrlSafeEncodedUsernameHash);
      return Optional.ofNullable(aci);
    } catch (IOException e) {
      return Optional.empty();
    }
  }

  public static String generateLink(String username) {
    String base64 = Base64UrlSafe.encodeBytesWithoutPadding(username.getBytes(StandardCharsets.UTF_8));

    return BASE_URL + base64;
  }

  /**
   * Parses the username from a link if possible, otherwise null.
   */
  public static @Nullable String parseLink(String url) {
    Matcher matcher = URL_PATTERN.matcher(url);
    if (!matcher.matches()) {
      return null;
    }

    String base64 = matcher.group(2);
    if (base64 == null) {
      return null;
    }

    try {
      return new String(Base64.decodeWithoutPadding(base64));
    } catch (IOException e) {
      return null;
    }
  }

  public enum InvalidReason {
    TOO_SHORT, TOO_LONG, INVALID_CHARACTERS, STARTS_WITH_NUMBER
  }
}

File: app/src/main/java/org/thoughtcrime/securesms/util/UsernameUtil.kt
package org.thoughtcrime.securesms.util

import org.signal.core.util.logging.Log
import org.signal.libsignal.usernames.BadDiscriminatorCharacterException
import org.signal.libsignal.usernames.BadNicknameCharacterException
import org.signal.libsignal.usernames.BaseUsernameException
import org.signal.libsignal.usernames.CannotBeEmptyException
import org.signal.libsignal.usernames.CannotStartWithDigitException
import org.signal.libsignal.usernames.DiscriminatorCannotBeEmptyException
import org.signal.libsignal.usernames.DiscriminatorCannotBeSingleDigitException
import org.signal.libsignal.usernames.DiscriminatorCannotBeZeroException
import org.signal.libsignal.usernames.DiscriminatorCannotHaveLeadingZerosException
import org.signal.libsignal.usernames.DiscriminatorTooLargeException
import org.signal.libsignal.usernames.NicknameTooLongException
import org.signal.libsignal.usernames.NicknameTooShortException
import org.signal.libsignal.usernames.Username
import java.util.Locale
import java.util.regex.Pattern

object UsernameUtil {
  private val TAG = Log.tag(UsernameUtil::class.java)
  const val MIN_NICKNAME_LENGTH = 3
  const val MAX_NICKNAME_LENGTH = 32
  const val MIN_DISCRIMINATOR_LENGTH = 2
  const val MAX_DISCRIMINATOR_LENGTH = 9
  private val FULL_PATTERN = Pattern.compile(String.format(Locale.US, "^[a-zA-Z_][a-zA-Z0-9_]{%d,%d}$", MIN_NICKNAME_LENGTH - 1, MAX_NICKNAME_LENGTH - 1), Pattern.CASE_INSENSITIVE)
  private val DIGIT_START_PATTERN = Pattern.compile("^[0-9].*$")
  private const val BASE_URL_SCHEMELESS = "signal.me/#eu/"
  private const val BASE_URL = "https://$BASE_URL_SCHEMELESS"

  private val SEARCH_PATTERN = Pattern.compile(
    String.format(
      Locale.US,
      "^@?[a-zA-Z_][a-zA-Z0-9_]{%d,%d}(.[0-9]+)?$",
      MIN_NICKNAME_LENGTH - 1,
      MAX_NICKNAME_LENGTH - 1,
      Pattern.CASE_INSENSITIVE
    )
  )

  @JvmStatic
  fun isValidUsernameForSearch(value: String): Boolean {
    return value.isNotEmpty() && SEARCH_PATTERN.matcher(value).matches()
  }

  @JvmStatic
  fun sanitizeUsernameFromSearch(value: String): String {
    return value.replace("[^a-zA-Z0-9_.]".toRegex(), "")
  }

  @JvmStatic
  fun checkNickname(value: String?): InvalidReason? {
    if (value == null) {
      return InvalidReason.TOO_SHORT
    }

    return try {
      // We only want to check the nickname, so we pass in a known-valid discriminator
      Username.fromParts(value, "01", MIN_NICKNAME_LENGTH, MAX_NICKNAME_LENGTH)
      null
    } catch (e: BadNicknameCharacterException) {
      InvalidReason.INVALID_CHARACTERS
    } catch (e: CannotBeEmptyException) {
      InvalidReason.TOO_SHORT
    } catch (e: CannotStartWithDigitException) {
      InvalidReason.STARTS_WITH_NUMBER
    } catch (e: NicknameTooLongException) {
      InvalidReason.TOO_LONG
    } catch (e: NicknameTooShortException) {
      InvalidReason.TOO_SHORT
    } catch (e: BaseUsernameException) {
      Log.w(TAG, "Unhandled verification exception!", e)
      InvalidReason.INVALID_CHARACTERS
    }
  }

  fun checkDiscriminator(value: String?): InvalidReason? {
    if (value == null) {
      return null
    }

    if (value.length < MIN_DISCRIMINATOR_LENGTH) {
      return InvalidReason.TOO_SHORT
    }

    if (value.length > MAX_DISCRIMINATOR_LENGTH) {
      return InvalidReason.TOO_LONG
    }

    return try {
      // We only want to check the discriminator, so we pass in a known-valid nickname
      Username.fromParts("spiderman", value, MIN_NICKNAME_LENGTH, MAX_NICKNAME_LENGTH)
      null
    } catch (e: BadDiscriminatorCharacterException) {
      InvalidReason.INVALID_CHARACTERS
    } catch (e: DiscriminatorCannotBeEmptyException) {
      InvalidReason.TOO_SHORT
    } catch (e: DiscriminatorCannotBeSingleDigitException) {
      InvalidReason.TOO_SHORT
    } catch (e: DiscriminatorCannotBeZeroException) {
      if (value.length < 2) {
        InvalidReason.TOO_SHORT
      } else if (value == "00") {
        InvalidReason.INVALID_NUMBER_00
      } else {
        InvalidReason.INVALID_NUMBER_PREFIX_0
      }
    } catch (e: DiscriminatorCannotHaveLeadingZerosException) {
      if (value.length < 2) {
        InvalidReason.TOO_SHORT
      } else if (value == "00") {
        InvalidReason.INVALID_NUMBER_00
      } else {
        InvalidReason.INVALID_NUMBER_PREFIX_0
      }
    } catch (e: DiscriminatorTooLargeException) {
      InvalidReason.TOO_LONG
    } catch (e: BaseUsernameException) {
      Log.w(TAG, "Unhandled verification exception!", e)
      InvalidReason.INVALID_CHARACTERS
    }
  }

  enum class InvalidReason {
    TOO_SHORT,
    TOO_LONG,
    INVALID_CHARACTERS,
    STARTS_WITH_NUMBER,
    INVALID_NUMBER,
    INVALID_NUMBER_00,
    INVALID_NUMBER_PREFIX_0
  }
}
